### PR TITLE
KAFKA-3311: Prepare internal source topics before calling partition grouper

### PR DESCRIPTION
**Pull request by guozhangwang**

### CODE REVIEW

**Comment by guozhangwang** [comment]
Ping @ymatsuda for reviews:

1) For all internal source topics, pre-validate the topic with the number of partitions to:
    a. number of partitions to the co-partition topics if there is any.
    b. maximum number of partitions to the depending sub-topologies (i.e. whose sink nodes include this internal source topic).

2) After calling partition grouper, re-validate the changelog and internal source topics to the maximum number of tasks.partitionId + 1.

3) move the co-partition checking logic to StreamsPartitionAssignor.

**Comment by ymatsuda** [comment]
What happens if auto.create.topics.enable=false?

**Comment by guozhangwang** [comment]
@ymatsuda Right now we have two rounds for create topics:

1) before calling partition grouper, create topics for internal source topics according to the logic above;

2) after calling partition grouper, create / update topics for changelog and internal source topics.

So we do not need to rely on broker side `auto.create.topics.enable` config.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java [hunk 1]
```diff
@@ -339,6 +403,43 @@ public void onAssignment(Assignment assignment) {
...
+        for (Set<String> copartitionGroup : copartitionGroups) {
+            ensureCopartitioning(copartitionGroup, internalTopics, metadata);
+        }
+    }
+
+    private void ensureCopartitioning(Set<String> copartitionGroup, Set<String> internalTopics, Cluster metadata) {
+        int numPartitions = -1;
+
+        for (String topic : copartitionGroup) {
+            List<PartitionInfo> infos = metadata.partitionsForTopic(topic);
```
Comment (ymatsuda) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java]: It seems you can move this line after line422.

Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java]: Ack, will fix.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java [hunk 2]
```diff
@@ -182,11 +186,71 @@ public Subscription subscription(Set<String> topics) {
...
+                do {
+                    partitions = streamThread.restoreConsumer.partitionsFor(topic);
+                } while (partitions == null || partitions.size() != numPartitions);
+            }
+
+            log.info("Completed validating internal source topics in partition assignor.");
         }
+
+        // get the tasks as partition groups from the partition grouper
         Map<TaskId, Set<TopicPartition>> partitionsForTask = streamThread.partitionGrouper.partitionGroups(sourceTopicGroups, metadata);
```
Comment (ymatsuda) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java:256]: Is `metadata` updated when internal topics are created?

Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java:256]: Good point. Will fix.

**Comment by guozhangwang** [comment]
@ymatsuda addressed comments.

**Comment by ymatsuda** [comment]
LGTM

**Comment by junrao** [comment]
Thanks for the patch. LGTM

**Pull request closed without merging**